### ISSUE TITLE
Unable to run IDLE without write-access to home directory

### ISSUE THREAD

**Issue by 2bbeb07f-5ba2-4529-bbf7-b5efbb677e3c** [python/cpython#52478]
BPO | [8231](https://bugs.python.org/issue8231)
--- | :---
Nosy | @terryjreedy, @ned-deily, @serwy, @asvetlov, @roseman, @mlouielu
PRs | <li>python/cpython#2629</li><li>python/cpython#2631</li>
Dependencies | <li>bpo-25507: IDLE: user code &#x27;import tkinter; tkinter.font&#x27; should fail</li><li>bpo-27534: IDLE: Reduce number and time for user process imports</li>
Files | <li>[error.png](https://bugs.python.org/file16649/error.png "Uploaded as image/x-png at 2010-03-25.15:39:26 by cane"): Error message</li><li>[@cfgdir.diff](https://bugs.python.org/file40872/%40cfgdir.diff "Uploaded as text/plain at 2015-10-28.06:36:01 by @terryjreedy"): Use temp dir as backup user config dir.</li><li>[@cfgdir2.diff](https://bugs.python.org/file40893/%40cfgdir2.diff "Uploaded as text/plain at 2015-10-29.08:59:56 by @terryjreedy")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/terryjreedy'
closed_at = None
created_at = <Date 2010-03-25.15:39:27.987>
labels = ['expert-IDLE', 'type-feature', '3.7']
title = 'Unable to run IDLE without write-access to home directory'
updated_at = <Date 2017-12-30.20:56:29.287>
user = 'https://bugs.python.org/cane'
```

bugs.python.org fields:
```python
activity = <Date 2017-12-30.20:56:29.287>
actor = 'ned.deily'
assignee = 'terry.reedy'
closed = False
closed_date = None
closer = None
components = ['IDLE']
creation = <Date 2010-03-25.15:39:27.987>
creator = 'cane'
dependencies = ['25507', '27534']
files = ['16649', '40872', '40893']
hgrepos = []
issue_num = 8231
keywords = ['patch']
message_count = 34.0
messages = ['101708', '101716', '156771', '156773', '156774', '188115', '222687', '222694', '222709', '222716', '252042', '253404', '253413', '253414', '253561', '253575', '253588', '253604', '253637', '253668', '253681', '253714', '253719', '297255', '297914', '297915', '297923', '297926', '297976', '298023', '301835', '309247', '309250', '309257']
nosy_count = 9.0
nosy_names = ['terry.reedy', 'ned.deily', 'roger.serwy', 'asvetlov', 'markroseman', 'cane', 'John Gray', 'tealduck', 'louielu']
pr_nums = ['2629', '2631']
priority = 'normal'
resolution = None
stage = 'needs patch'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue8231'
versions = ['Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by terryjreedy** [comment]
Problems with patching 2.7 are no longer relevant.

To test, we should refactor config so that the attempt to find and access $HOME and .idlerc are isolated in a function that can be mocked to simulate various problems.

```py
def get_rc():
   """Return a directory path that is readable and writable.
```


   If not possible, return an error indicator.  \<to be determined\>
   """

Testing such a function is a different issue, but I would just reuse existing code.

**Comment by terryjreedy** [comment]
> GetUserCfgDir() is called in three places, (buried in one of Mark's posts).  On the fact of it, this seems like something that should be fixed.

**Comment by terryjreedy** [comment]
My second patch did fix that.  I think I will extract that part immediately.

**Comment by terryjreedy** [comment]
New changeset 223c7e70e48eb6eed4aab3906fbe32b098faafe3 by terryjreedy in branch 'master':
bpo-8231: Call idlelib.IdleConf.GetUserCfgDir only once. (bpo-2629)
https://github.com/python/cpython/commit/223c7e70e48eb6eed4aab3906fbe32b098faafe3

**Comment by terryjreedy** [comment]
New changeset 552f26680d3806df7c27dd7161fd7d57ac815f78 by terryjreedy in branch '3.6':
[3.6] bpo-8231: Call idlelib.IdleConf.GetUserCfgDir only once. (GH-2629) (bpo-2631)
https://github.com/python/cpython/commit/552f26680d3806df7c27dd7161fd7d57ac815f78

**Comment by terryjreedy** [comment]
bpo-27534 is about reducing the imports into the user runcode process. Without rechecking the reason for each import, I think it possible that this might result in config not being indirectly imported.

**Comment by 9da18a29-a07d-4828-a801-24cc911246de** [comment]
How about only taking warning when not able to create dir at GetUserCfgDir(), then take the permission handler in other place?

e.g. when user trying to save the config in bad dir, pop-out a dialog to tell it is permission problem or dir not eixsts...etc.

**Comment by terryjreedy** [comment]
I closed bpo-30918 as a duplicate of this.  It has full 'set' and expanduser info.

**Comment by erlend-aasland** [comment]
IDLE now prints a user friendly warning when trying to start it with a read-only home directory; see https://github.com/python/cpython/issues/52478#issuecomment-1093496334.

Suggesting to close this, as the OP problem seems resolved. If other improvements are needed, I suggest to create targeted, narrow issues for each of these.

**Comment by terryjreedy** [comment]
Given Ned's comment and that improvements were made both here and on two other issues, agreed.
